### ISSUE TITLE
"Package name is too similar to another package" issue

### ISSUE THREAD

**Issue by jblough** [dart-lang/pub-dartlang-dart#2145]
Based on reading the encouragement by the Dart package community, I was trying to upload my package as a new package (tincan) instead of trying to rename the existing package (tin_can).  This is failing with the error "Package name is too similar to another package".  What is the recommended way to continue development on the package as "tincan" and still publish to pub.dartlang.org?  Thank you!

**Comment by jonasfj** [comment]
We forbid packages with similar names to prevent typo squatting... Our heuristic isn't perfect.. 

In your case, since you uploaded the other package and it was uploaded recently, it might be reasonable to file a request here to have the original package deleted.. then you would be able to upload the other package.

Note. I would verify the deletion request... So it might take some time...

**Comment by jblough** [comment]
Thank you!  Should I file the deletion request on this issue or create another issue for that and reference this issue?

**Comment by jonasfj** [comment]
Just comment here is fine :)

**Comment by jblough** [comment]
Okay thanks!  Please delete the tin_can package.

**Comment by jonasfj** [comment]
request confirmed out of band.

**Comment by jonasfj** [comment]
package removed.

**Comment by thosakwe** [comment]
Just attempted to publish a package named `dart_up`, and was met with this message. It indeed turns out that there's a package named `dartup` (it hasn't been updated since 2015, but for all I know there could still be many people using it).

Is the best idea to just change the name of my own package, then?

**Comment by isoos** [comment]
> Is the best idea to just change the name of my own package, then?

You can also contact the author of the abandoned package and ask them if they are willing to give it up to you. They can invite you to become an uploader of the package, and then you can remove them from it, and upload new versions. (A quick search revealed that the author seems to be active in other github repos).

We have a few ideas how to help this process in the long term, but right now either this or rename.

**Comment by thosakwe** [comment]
Sounds good, thanks.

**Comment by topperspal** [comment]
2-3-year-old packages spoiling the latest packages names. I want to upload a package named `forms` but the command throws the following error-
```cmd
Package name `forms` is too similar to another active package: `form` (https://pub.dev/packages/form).
```

I need a small and concise name so that I can use it in the file extension `.forms.dart` for the form template file and `.g.dart` for generated form handler file.

Please help me publish the package.

**Comment by isoos** [comment]
@topperspal: First of all, nothing prevents you to use longer package names and short (but maybe configurable) suffixes for your special files. The chance of name collision is small, and if you make it configurable, it won't matter much.

There is an alternate route: the package author of `form` has a contact email on their GitHub page. You could contact them, describe your case, ask for the plans with the package, and if they don't want continue it, you could ask for transferring it to you. Please keep in mind that this is a voluntary decision, they are not obliged to do anything about it.

**Comment by MayurSMahajan** [comment]
How do I change the name of my package?

**Comment by isoos** [comment]
@MayurSMahajan: https://dart.dev/tools/pub/pubspec#name

**Comment by isoos** [comment]
@MayurSMahajan:

"Pub.dev facilitates the sharing via publication of Dart packages. It is central to this service that consumers of packages can trust that their dependencies do not suddenly disappear. Thus, once a package has been published it cannot be unpublished or deleted. This applies to all versions of a published package. If something has been published that is no longer relevant or maintained, as a publisher, you can mark the package as discontinued, which will make the package disappear from searches (but keep it available to those that already depend on it)."

https://pub.dev/policy

In short: once published, the package name keeps being the same. You may publish the same package under a different name, but the original one remains.

**Comment by MayurSMahajan** [comment]
Thanks @isoos!

**Comment by medz** [comment]
What a bad decision, no package management platform prohibits something like this. However, there are many packages on pub.dev that have not been maintained for more than 5 years.

For example, I am developing the Prisma dart client for Dart, and this behavior is officially supported by Prisma ORM. Last year, I tried to contact pub.dev to submit a deletion request or transfer request. And actively contact the existing `package:prisma` authors to get any replies.

When I develop a lot of infrastructure and publish packages, such as the `FormData` package, the existing formdara name is not used, but formdata is prohibited from being published. The reason is that there are two packages similar to formdata.

There is no way to protest this behavior on pub.dev, which is a really bad experience.

**Comment by topperspal** [comment]
same issue with me. I created a form handler lib and wanted to name it forms but it says the name forms is ambiguous because there are already two packages published named forme and formz. this is really bad

**Comment by medz** [comment]
Due to the publishing restrictions on this name, I now have to write myself out a rough development plan for a year and then seize the name I want to use in advance. Regardless of whether these open source projects will be released in the end, I will mark the ones that have not been released as unused and attach my contact information.

Yes, I know this is against pub.dev policy rules, but this is not my intention, this is caused by "disallowing similar package names" and not supporting "package names containing namespaces". If I don't take it early, every time I finish developing a package, I have to rename it and try to publish it to see if it complies with the rules.

It can be seen from my actual operation and the large number of packages on pub now that this rule not only does not effectively curb the problem of malicious occupation, but on the contrary, makes developers more active in occupying the package names they want.

The simplest example is the Prisma client I developed for Dart. Since I couldn't contact the author of `package:prisma` at all (the package has been abandoned and is no longer maintained), I once sent an email to `support@pub.dev ` Asking for help, but in the end I didn't get any reply. In desperation, I started to occupy the common name `orm`.

> PS: Starting this year, my thinking began to change, and I started using meaningless names as package names, such as `wow`/`ovo`

In my opinion, this is a helpless move. It takes a certain amount of time for a developer to develop a package. During this period, someone may have occupied the package name, but its content is very different from the package name.

A similar thing happened with NPM in the past, but npm did not choose to ban similar package names, but introduced org rules, that is, something like `@vuejs/core`, and achieved good benefits.

**Comment by isoos** [comment]
> Due to the publishing restrictions on this name, I now have to write myself out a rough development plan for a year and then seize the name I want to use in advance.

While having a common name helps a bit in package discovery, the quality of the code, the availability and the track record of the authors are usually far more important than the name. For example, suppose if 3 teams start to develop an ORM with 3 very different API approach and philosophy (not an unlikely scenario given a few recent reddit threads), which should be awarded to use the `package:orm` name? In a flat namespace, there is no perfect arbitration to decide on this. The first one gets the `orm`, the others will chose a different name. Their discoverability won't be that harder than much different than having three `randomorg/orm` packages in search results.

Over the years I have seen people doing similar things with big plans about their future packages. One such notorious example is https://pub.dev/packages/mysql which I remember because I've contacted the author to give up the name to another, actively developed mysql client. They insisted that they have solid development plans. Life changes, people change, goals change, and they did not release any further versions. On the other hand, based on the popularity numbers, both `mysql1` and `mysql_client` is in use, so people may not need the simplest name for such packages.

**Comment by medz** [comment]
@isoos I agree with you, my planned early occupation happened before this year.

This year I'm focusing more on meaningless package names. I'll give my package a random name, and then I won't develop content in the lib until the first version is released. Instead, I create a new dev directory (this behavior is to prevent the dart extension from automatically importing code through package:xxx). After I complete the development, I will rename dev to the lib directory. Subsequent modifications will make no sense, and For a name that has nothing to do with the package itself but is easy to remember, go to pub.dev to check whether it exists. If it exists, I will think of a new one (under English grammar, naming a meaningless and easy-to-remember name is too simple) and then publish this package.
